### DISCUSSION TITLE
Lesson 9: Why do we need to emit an event again manually in test script?

### DISCUSSION THREAD

**Discussion by TreapGoGo** [Cyfrin/foundry-full-course-cu#645] [Q&A] [answered]
In `RaffleTest.t.sol` we have

```solidity
function testEmitsEventOnEntrance() public {
        vm.prank(PLAYER);
        vm.expectEmit(true, false, false, false, address(raffle));
        emit EnteredRaffle(PLAYER);
        raffle.enterRaffle{value: entranceFee}();
    }
```

It really confuses me that we still need to write `emit EnteredRaffle(PLAYER);` just here, because what we want to test is the event in `Raffle.sol` instead of `RaffleTest.s.sol`.

When `raffle.enterRaffle{value: entranceFee}();` is running, an `EnteredRaffle` event will be emitted. Isn't this the expected event?

Any help would be appreciated :)

**Comment by theirrationalone** [comment, marked as answer]
Because it's detached from contract storage it's only for logging the transaction data that was recorded & stored into logs when a transaction went through.... 

Remember when Patrick was teching us "we can store data into six places"... 

1. Code
2. Stack
3. Logs
4. Memory
5. Calldata
6. Storage

So in the tests when we just only have an instance of a deployed contract. And that contract doesn't have any information about events because events doesn't get stored into contact's storage. 

So that's why we need to mimic events as we expect that contract will emit/trigger on certain event(any outgoing tx. Whose data worth to be traced which can help us to confirm some consequences). 

The thing to be noted here, Is that... 

1. We cannot index more than 3 data elements. 
2. indexed typed must be same as of contract's events data indexed type in very same order. 
3. We really don't care about index in testing we can toggle whether we want to check element indexed at specific positions Or not, By making relative positions true Or false. 

In context of making DAPPs & WEB3 Websites You will find **EVENTS** plays most important roles to account, monitor, structure, enhancing visual representation, of data on DAPPs & WEB3 websites. 

**Even though, Having all the data monitored by web2 protocols, Nobody  can modify the state of blockchain.**

**Reply by Niferu** [reply]
Yes and that is why we need to recreate this event in our test to mirror real event and get info from logs for example.
